### DISCUSSION TITLE
Fabrication v2: Alpha release

### DISCUSSION THREAD

**Discussion by MrTJP** [MrTJP/ProjectRed#1730] [Announcements]
Alpha release is dropping now. Here's some things to help you get started.

## Downloads
* [project-red-4.13.0-alpha-6.zip](https://github.com/MrTJP/ProjectRed/files/8850672/project-red-4.13.0-alpha-6.zip)

## Getting started
This release of Fabrication contains these blocks and items:
* IC Workbench block
* IC Blueprint item
* Fabricated Gate block

### Overview
1. Place an IC Workbench block
2. Place down an IC Blueprint item onto the workbench by right-clicking
3. Open the workbench GUI and edit your circuit
4. Take the blueprint off the workbench via sneak-right-click with empty hand. Blueprint now has a circuit inside.
5. Right-click with the blueprint to obtain a Fabricated Gate item
6. Place gate and use your circuit in-world

### Layout and controls
* Pan around with `W/A/S/D`
* Zoom in/out with `scroll`
* Layer up/down with `arrow up`/`arrow down` (not useful yet, as vertical wires are to come later)
* Select a tool with the tabs on the right side
* Ignore tabs on left side for now (not implemented, they are empty)

### Editor tools
* `Interact tool` is used to configure gates and other tiles. Tiles that can be configured will display blue regions that can be clicked (zoom all the way in and hover over the tile to see them).
* `Eraser tool` is used to delete tiles. Can be dragged to select a box
* `Gate tool` is used to place gates. `IO Gate` tiles can only be placed on the gold edge, and regular gates only inside the body of the circuit
* `Wire tool` is used to place redstone wires and bundled cables

### Gate IO connections
* Fabricated gate blocks will show bundled connections on whatever edges that contained IO gates (only bundled for now)
* Each IO gate can have the `direction` and `colour` configured via the `Interact` tool
* For input gates, you can feed in a signal and see the circuit simulate in real time. Use the interact tool to toggle that channel's input.

## Known Issues and Caveats
* This is purely creative-only for this first release, intended to allow you to mess around with the circuit compiler.
* Some gates are still missing. Check the project board for progress on adding the rest.
* RS Latch render connected red alloy wires strangely due to them having a side that is both input and output. They will most likely be changed to not back-feed into the input sides because it doesn't really make sense in this context.
* There is no GUI layer to show warnings and improper hookups. So you will be able to place incorrectly built circuits and may end up thinking they are bugged, while in reality, they are just made incorrectly. Some things to keep in mind:
    * 2 gates cannot both drive a single input of another gate. The signals _must_ be OR'ed together, unlike regular Minecraft redstone, where the wire implicitly OR's multiple input signals in the wire itself. On the contrary, a single output _can_ drive multiple inputs.
    * Multiple IO gates driving the same colour can be placed on the same edge, and even configured to different directions. While this may "work" in some cases, it also opens the door for self-powering loops. Having the entire edge be a single direction will be enforced eventually.
* Panning/scrolling animation speed is unintentionally linked to FPS. Will be fixed soon but for now, if you are working with low fps, you will find it slightly slow to pan around. You can use zoom for assistance, as zoom is positional.

## Development
You can keep up to date with development by following the `feature/fabrication-v2` branch and the [Fabrication v2 project board](https://github.com/users/MrTJP/projects/4/views/6). If you'd like to build the latest commit yourself, simply check out that branch and run `./gradlew build`. I'll occasionally push `feature/fabrication-v2` -> `publish/1.16/alpha` to trigger an auto build and maven upload, then post the artifacts above on this discussion.

## Feedback
Please keep feedback on this thread only. No need to open issues until things are more stable and ready for Curseforge upload.

**Comment by Rayzeq** [comment]
Hey!
I tested fabrication beta today (yes, I'm a bit late), and I think the new interface is really good! However, there are still some issues, here's a quick list:

- There is a `fabricated_gate` in the creative menu that shouldn't be there.
- I find zoomming a bit too slow (at 60 fps)
- I can't pan, even though your post suggests it's implemented (I tried left, middle and right click, with and without control and shift)
- I accidentally discovered how to change layers, I think the controls should be written somewhere.
- Everything disappears when zooming out too much, this is normal but I think the clipping plane should be farther
- Everything also disappears when zooming in, but does not reappear when zooming out, you have to leave the workbench.
- It quickly becomes laggy when there are too many things, visual and zooming does not lag, but for example changing io direction make the game freeze for a short time
- The default grid size should be 18x18, so you can put all 16 io colors on one side.
- In the "wires" tab, the red alloy and the bundled uncolored wires should have a square selection box rather than a rectangle that takes up all the available space.
- I guess this is not implemented yet, but there is no unbundled IO.

**Reply by MrTJP** [reply]
1. Acknowledged. Will deal with that at some point
2. Just fixed via 54018437292440d6eac3a351d8e8dbaff4a83c92
3. See above. Panning is done with W/A/S/D
4. They will be clearer once changing layers actually matters
5. Yup, adding a zoom out/in limit
6. Same as above
7. Sounds like slow-down when you do anything to trigger a circuit change. Everything gets re-compiled. But it should happen gradually over the next X ticks, not all at once. Weird that your entire game freezes. Do you have a picture of your circuit or something?
8. Considering this. Size probably should refer to internal size without the IO edges
9. Mostly temporary. Gets the job done for now but will likely change
10. Yup, others to be implemented

**Reply by Rayzeq** [reply]
Okay so I tested the 4.13.0-alpha6 version, and here are the issues I found:

- I don't know if the zoom fix has been applied in this version but I still think it's a bit too slow
- I think panning should be done with middle-click + drag, because using wasd is not very practical (and there is no intuitive behavior with wasd, it could move the view, as currently, or it could move the circuit)
- I made this circuit by randomly placing things:
![Capture d’écran du 2022-10-28 17-05-28](https://user-images.githubusercontent.com/26172375/198673264-91825b31-6702-406d-82d4-a94d1f4a2e6b.png)
And now if I try to change anything (e.g the direction of any io gate), the game don't freeze but it takes something like 2 or 3 seconds before it shows up

**Reply by MrTJP** [reply]
Uploading a new one today with a bunch of fixes, but to answer your questions:
1. Yes, zoom is now tied to absolute time rather than frame intervals
2. Pretty much all recent games I've played use WASD to move around (Factorio, etc). That's where the inspiration came from. I may change this in the future but for now, it will stay as is. The entire UI has been built around this control scheme.
3. Also fixed. Now you have to go to the compile tab and re-build the changes, and it will not auto-build every time you change something (will be configurable. Server admins will probably want to disable auto-rebuild). It is also a multi-tick operation instead of single-tick, so you won't get the freeze.
4. Btw, the weird rendering of the wires above is because you're breaking the single-driver rule (same as previous versions of fabrication). An output may drive multiple inputs, but any input must only have 1 input. Its not obvious yet because no "issues" overlay.

**Comment by MCPlayer576** [comment]
Hello! The Mod is working well. But I wanted to know if you are going to add the IC Printer.
